src/test/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPluginTest.java

```diff
@@ -0,0 +1,231 @@
+
+    this.plugin = new AWSSecretsManagerPlugin(
+        proxy,
+        propertySet,
+        nextPlugin,
+        logger,
+        mockSecretsManagerClient,
+        mockGetValueRequest);
+  }
+
+  @AfterEach
+  void cleanUp() throws Exception {
+    closeable.close();
+    AWSSecretsManagerPlugin.SECRET_CACHE.clear();
+  }
+
+  /**
+   * The plugin will successfully open a connection with a cached secret.
+   */
+  @Test
+  public void testConnectWithCachedSecrets() throws SQLException {
+    // Add initial cached secret to be used for a connection
+    AWSSecretsManagerPlugin.SECRET_CACHE.put(SECRET_CACHE_KEY, SECRET);
+
+    this.plugin.openInitialConnection(this.connectionUrl);
+
+    assertEquals(1, AWSSecretsManagerPlugin.SECRET_CACHE.size());
+    verify(this.mockSecretsManagerClient, never()).getSecretValue(this.mockGetValueRequest);
+    verify(this.nextPlugin).openInitialConnection(this.captor.capture());
+    final List<ConnectionUrl> connectionUrls = this.captor.getAllValues();
+    final Map<String, String> connectionProperties = connectionUrls.get(0).getOriginalProperties();
+    assertEquals(TEST_USERNAME, connectionProperties.get("user"));
+    assertEquals(TEST_PASSWORD, connectionProperties.get("password"));
+  }
+
+  /**
+   * The plugin will attempt to open a connection with a cached secret, but it will fail with a generic SQL exception.
+   * In this case, the plugin will rethrow the error back to the user.
+   */
+  @Test
+  public void testFailedInitialConnectionWithGenericError() throws SQLException {
+    final SQLException failedFirstConnectionGenericException = new SQLException(TEST_SQL_ERROR, SQL_STATE_FAIL_GENERIC);
+    doThrow(failedFirstConnectionGenericException).when(this.nextPlugin).openInitialConnection(any(ConnectionUrl.class));
+
+    final SQLException connectionFailedException = assertThrows(SQLException.class, () -> this.plugin.openInitialConnection(this.connectionUrl));
+
+    assertEquals(TEST_SQL_ERROR, connectionFailedException.getMessage());
+    assertEquals(0, AWSSecretsManagerPlugin.SECRET_CACHE.size());
+    verify(this.mockSecretsManagerClient, never()).getSecretValue(this.mockGetValueRequest);
+    verify(this.nextPlugin).openInitialConnection(this.captor.capture());
+    final List<ConnectionUrl> connectionUrls = this.captor.getAllValues();
+    final Map<String, String> connectionProperties = connectionUrls.get(0).getOriginalProperties();
+    assertNull(connectionProperties.get("user"));
+    assertNull(connectionProperties.get("password"));
+  }
+
+  /**
+   * The plugin will attempt to open a connection with a cached secret, but it will fail with an access error. In
+   * this case, the plugin will fetch the secret and will retry the connection.
+   */
+  @Test
+  public void testConnectWithNewSecrets() throws SQLException {
+    // Fail initial connection attempt so secrets will be retrieved. Second attempt should be successful
```

Comment:
Missing period at end of comment on line 163.

Reply:
not sure I've ever worried about periods in comments, unless it was a multiple sentences.